### PR TITLE
feat(eip7702): Impl newest version of EIP 

### PR DESCRIPTION
**Pull request by rakita**
Implementation of EIP-7702,

* Added StateLoad types to Host to structure is_cold field.
* Added new handler or eip7702 load
* Split refund logic from `last_frame_handler` to its own handler `post_exec.refund` it is more aligned with handling those outside of execution. Optimism handler updated
* Added last test suite for bls and eip7702 EIPs.

### CODE REVIEW

File: crates/primitives/src/eip7702/bytecode.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
+use crate::{bytes, Address, Bytes};
+
+/// EIP Version Magic in u16 form.
+pub const EIP7702_MAGIC: u16 = 0xEF01;
+
+/// EOF magic number in array form.
```
Comment (roynalnaruto) [inline crates/primitives/src/eip7702/bytecode.rs:6]:
```suggestion
/// EIP-7702 magic number in array form.
```

File: crates/primitives/src/eip7702/bytecode.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+pub static EIP7702_MAGIC_BYTES: Bytes = bytes!("ef01");
```
Comment (roynalnaruto) [inline crates/primitives/src/eip7702/bytecode.rs:7]:
```suggestion
pub static EIP7702_MAGIC_BYTES: Bytes = bytes!("ef0100");
```

As of the [latest commit](https://github.com/ethereum/EIPs/blob/52de6be37bfb7c134d0a2280f3d5b3d2433dc572/EIPS/eip-7702.md), this should be `0xef0100` ?

File: crates/primitives/src/eip7702/bytecode.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+pub struct Eip7702Bytecode {
+    pub delegated_address: Address,
+    pub raw: Bytes,
+}
+
+impl Eip7702Bytecode {
+    /// Creates a new EIP-7702 bytecode or returns None if the raw bytecode is invalid.
+    #[inline]
+    pub fn new(raw: Bytes) -> Option<Self> {
+        if raw.len() != 22 {
```
Comment (roynalnaruto) [inline crates/primitives/src/eip7702/bytecode.rs:20]:
```suggestion
        if raw.len() != 23 {
```

File: crates/primitives/src/eip7702/bytecode.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            return None;
+        }
+        if !raw.starts_with(&EIP7702_MAGIC_BYTES) {
+            return None;
+        }
+        Some(Self {
+            delegated_address: Address::new(raw[2..].try_into().unwrap()),
```
Comment (roynalnaruto) [inline crates/primitives/src/eip7702/bytecode.rs:27]:
```suggestion
            delegated_address: Address::new(raw[3..].try_into().unwrap()),
```

**Review by roynalnaruto — COMMENTED** [review]
As of the `master` branch's version of EIP-7702, the prefix would be `0xef0100`, and according to that I have proposed some edits. Sorry I know this is a WIP PR, simply informing as I was referring the EIP and this implementation :)

**Comment by rakita** [comment]
> As of the `master` branch's version of EIP-7702, the prefix would be `0xef0100`, and according to that I have proposed some edits. Sorry I know this is a WIP PR, simply informing as I was referring the EIP and this implementation :)

Oh, i missed this, thank you for taking a look!

File: crates/primitives/src/eip7702/bytecode.rs (hunk 2 above)
Comment (rakita) [reply crates/primitives/src/eip7702/bytecode.rs:7]: MAGIC filed is something similar to EOF magic, so I will leave this const as same, and add additional const `EIP7702_VERSION`

File: crates/primitives/src/bytecode.rs [hunk 5]
```diff
@@ -117,6 +138,10 @@ impl Bytecode {
     /// Returns a reference to the bytecode.
     ///
     /// In case of EOF this will be the first code section.
+    ///
+    /// # Panics
+    ///   
+    /// Panics if Bytecode is EIp7702 that does not contain bytecode.
```
Comment (onbjerg) [inline crates/primitives/src/bytecode.rs:144]:
```suggestion
    /// Panics if Bytecode is Eip7702 that does not contain bytecode.
```

File: crates/primitives/src/bytecode.rs [hunk 6]
```diff
@@ -18,6 +20,8 @@ pub enum Bytecode {
     LegacyAnalyzed(LegacyAnalyzedBytecode),
     /// Ethereum Object Format
     Eof(Arc<Eof>),
+    /// EIP-7702 delegated bytecode
+    Eip7702s(Eip7702Bytecode),
```
Comment (onbjerg) [inline crates/primitives/src/bytecode.rs:24]:
```suggestion
    Eip7702(Eip7702Bytecode),
```

seems like a typo?

File: crates/primitives/src/constants.rs [hunk 7]
```diff
@@ -10,7 +10,7 @@ pub const BLOCK_HASH_HISTORY: usize = 256;
 
 /// EIP-2935: Serve historical block hashes from state
 ///
-/// Number of block hashes the EVM can access in the past (Prague).
+/// Number of block hashes the EVM –can access in the past (Prague).
```
Comment (onbjerg) [inline crates/primitives/src/constants.rs:13]:
```suggestion
/// Number of block hashes the EVM can access in the past (Prague).
```

File: crates/primitives/src/eip7702.rs [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+pub use authorization_list::{
+    Authorization, AuthorizationList, RecoveredAuthorization, Signature, SignedAuthorization,
+};
+pub use bytecode::{Eip7702Bytecode, EIP7702_MAGIC, EIP7702_MAGIC_BYTES, EIP7702_VERSION};
+
+// Base cost of updating authorized account.
+pub const PER_AUTH_BASE_COST: u64 = 2500;
+
+/// Cost of creating authorized account that was previously empty.
+pub const PER_EMPTY_ACCOUNT_COST: u64 = 25000;
```
Comment (onbjerg) [inline crates/primitives/src/eip7702.rs:13]: these are in alloy_eips

File: crates/primitives/src/eip7702/authorization_list.rs [hunk 9]
```diff
@@ -34,6 +34,18 @@ impl AuthorizationList {
...
+    /// Returns true if the authorization list is valid.
+    /// TODO EIP-7702 switch to error type.
+    pub fn is_valid(&self) -> bool {
+        true
+        /*
+        match self {
+            Self::Signed(signed) => signed.iter().all(|signed| signed.is_valid()),
+            Self::Recovered(recovered) => recovered.iter().all(|recovered| recovered.is_valid()),
+        }
+        */
```
Comment (onbjerg) [inline crates/primitives/src/eip7702/authorization_list.rs:46]: need to impl this for alphanet

File: crates/primitives/src/eip7702/bytecode.rs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+pub struct Eip7702Bytecode {
+    pub delegated_address: Address,
+    pub version: u8,
+    pub raw: Bytes,
+}
+
+impl Eip7702Bytecode {
+    /// Creates a new EIP-7702 bytecode or returns None if the raw bytecode is invalid.
+    #[inline]
+    pub fn new(raw: Bytes) -> Option<Self> {
```
Comment (onbjerg) [inline crates/primitives/src/eip7702/bytecode.rs:27]: nit: prob better to have `new` take address and do a `new_raw` instead (i.e. flip `new` and `new_address`)

File: crates/primitives/src/eip7702/bytecode.rs [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    pub fn address(&self) -> Address {
+        self.delegated_address
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
```
Comment (onbjerg) [inline crates/primitives/src/eip7702/bytecode.rs:76]: should we add a test for len > 23, wrong version and wrong magic?

File: crates/revm/src/handler/mainnet/pre_execution.rs [hunk 12]
```diff
@@ -153,14 +85,90 @@ pub fn deduct_caller<SPEC: Spec, EXT, DB: Database>(
...
+
+        // 3. Verify the code of authority is either empty or already delegated.
+        if let Some(bytecode) = &authority_acc.info.code {
+            // if it is not empty or it is not eip7702
+            if !bytecode.is_empty() || !bytecode.is_eip7702() {
+                continue;
+            }
+        }
+
+        // 4. If nonce list item is length one, verify the nonce of authority is equal to nonce.
```
Comment (onbjerg) [inline crates/revm/src/handler/mainnet/pre_execution.rs:146]: nit: comment is outdated

File: crates/revm/src/handler/mainnet/pre_execution.rs [hunk 13]
```diff
@@ -153,14 +85,90 @@ pub fn deduct_caller<SPEC: Spec, EXT, DB: Database>(
...
+        // 6. Set the code of authority to be 0xef0100 || address. This is a delegation designation.
+        let bytecode = Bytecode::new_eip7702(authorization.address);
+        authority_acc.info.code_hash = bytecode.hash_slow();
+        authority_acc.info.code = Some(bytecode);
+
+        // 7. Increase the nonce of authority by one.
+        authority_acc.info.nonce = authority_acc.info.nonce.saturating_add(1);
+        authority_acc.mark_touch();
+
+        // TODO(EIP-7702) set contract to account.
```
Comment (onbjerg) [inline crates/revm/src/handler/mainnet/pre_execution.rs:167]: is this not done above?

**Review by onbjerg — COMMENTED** [review]
preliminary review

File: crates/primitives/src/eip7702/authorization_list.rs (hunk 9 above)
Comment (rakita) [reply crates/primitives/src/eip7702/authorization_list.rs:46]: Would not agree it is needed for alphanet. But for eth yes

File: crates/primitives/src/eip7702.rs (hunk 8 above)
Comment (rakita) [reply crates/primitives/src/eip7702.rs:13]: It is more redable if they are in same repo

File: crates/revm/src/handler/mainnet/pre_execution.rs (hunk 13 above)
Comment (rakita) [reply crates/revm/src/handler/mainnet/pre_execution.rs:167]: comment removed

File: crates/revm/src/handler/mainnet/pre_execution.rs (hunk 12 above)
Comment (rakita) [reply crates/revm/src/handler/mainnet/pre_execution.rs]: Updated the comment and enumerations

File: crates/primitives/src/eip7702/bytecode.rs (hunk 10 above)
Comment (rakita) [reply crates/primitives/src/eip7702/bytecode.rs]: It is better to switch them.

File: crates/revm/src/evm.rs [hunk 14]
```diff
@@ -387,3 +394,53 @@ impl<EXT, DB: Database> Evm<'_, EXT, DB> {
...
+
+    use super::*;
+    use crate::{
+        db::BenchmarkDB,
+        interpreter::opcode::{PUSH1, SSTORE},
+        primitives::{address, Authorization, Bytecode, RecoveredAuthorization, U256},
+    };
+
+    #[test]
+    fn sanity_eip7702_tx() {
```
Comment (rakita) [inline crates/revm/src/evm.rs:409]: @onbjerg @rkrasiuk sanity delegation test

File: crates/primitives/src/eip7702/bytecode.rs (hunk 11 above)
Comment (rakita) [reply crates/primitives/src/eip7702/bytecode.rs:103]: done

**Pull request merged**